daemon/api.go

```diff
@@ -2635,56 +2627,13 @@ func postApps(c *Command, r *http.Request, user *auth.UserState) Response {
 		return InternalError("no services found")
 	}
 
-	// the argv to call systemctl will need at most one entry per appInfo,
-	// plus one for "systemctl", one for the action, and sometimes one for
-	// an option. That's a maximum of 3+len(appInfos).
-	argv := make([]string, 2, 3+len(appInfos))
-	argv[0] = "systemctl"
-
-	argv[1] = inst.Action
-	switch inst.Action {
-	case "start":
-		if inst.Enable {
-			argv[1] = "enable"
-			argv = append(argv, "--now")
-		}
-	case "stop":
-		if inst.Disable {
-			argv[1] = "disable"
-			argv = append(argv, "--now")
-		}
-	case "restart":
-		if inst.Reload {
-			argv[1] = "reload-or-restart"
-		}
-	default:
-		return BadRequest("unknown action %q", inst.Action)
-	}
-
-	snapNames := make([]string, 0, len(appInfos))
-	lastName := ""
-	names := make([]string, len(appInfos))
-	for i, svc := range appInfos {
-		argv = append(argv, svc.ServiceName())
-		snapName := svc.Snap.Name()
-		names[i] = snapName + "." + svc.Name
-		if snapName != lastName {
-			snapNames = append(snapNames, snapName)
-			lastName = snapName
+	chg, err := servicectl.ServiceControl(st, appInfos, &inst)
+	if err != nil {
+		if _, ok := err.(servicectl.ServiceActionConflict); ok {
+			return InternalError(err.Error())
```

Comment:
An internal error is a bug in our logic: it should _never_ happen, but we failed somehow. A change conflict is a completely reasonable error that we need to report back, but that can happen very naturally.

Luckily, there's a good error type for conflicts.. it's called _Conflict._ :smile: